### ISSUE TITLE
Support for Vert.x

### ISSUE THREAD

**Issue by glamarre360** [stagemonitor/stagemonitor#253]
I'm trying to implement the request monitor in an async context where the request aren't all executed on the same thread. I managed to delay monitor stop of parent request and to set a hierarchy in the requests infos but i can't see that hierarchy on request analysis in kibana.

**Comment by felixbarny** [comment]
This is something I'm currently working on in the context of #118. See also the pull request #252. Are you executing your requests in a `ExecutorService`?

Closing as duplicate of #118 but feel free to comment here

**Comment by glamarre360** [comment]
The whole async is managed by rxJava observable I don't quite understand how to trace those calls with your new architecture.

**Comment by felixbarny** [comment]
I'm sick atm. I'll answer at the weekend. In short: the context has to be propagated to the executing thread.

**Comment by glamarre360** [comment]
For you to understand our specific architecure we use vert.x and an eventbus system. Which when a message is received a sub thread is openned doing a part of the work and this thread may send message on the eventbus meaning the work will continue on an other thread. But those two work thread aren't linked by nothing.

**Comment by felixbarny** [comment]
I see, so this is basically a feature request for Vert.x in disguise?

There are currently some ongoing discussions in the OpenTracing community on how to do this properly. See https://github.com/opentracing/specification/issues/23 and https://github.com/opentracing/opentracing-java/pull/111.

This is a concrete example on how the propagation could work for rxJava: https://github.com/smoketurner/dropwizard-zipkin/blob/master/zipkin-rx/src/main/java/com/smoketurner/dropwizard/zipkin/rx/BraveRxJavaSchedulersHook.java

While the new architecture makes it possible to support Vert.x, it will probably take a while until it gets implemented unless someone (like you maybe? 😉) provides a solid pull request.

Stagemonitor is currently using jaeger-context for span context propagation which provides support for creating context propagating ExecutorServices.

**Comment by glamarre360** [comment]
I have seen your test case for the executor service and I see you do a monitor stop directly after the async call. Won't that cause stagemonitor to report uncompleted request if the async job is a bit long?

**Comment by felixbarny** [comment]
I'm not quite sure I understand what you mean. The original request and the async execution are monitored separately (both are represented as a `Span` of the same `Trace`) and can be reported separately.

**Comment by glamarre360** [comment]
But then how does the hierarchy between the different requests is keeped. I mean how i see in the data that a request led to another and so on.

**Comment by felixbarny** [comment]
`TracingUtils.tracedExecutor` creates an `ExecutorService` which propagates the current Span context to the threads of the `ExecutorService`.

**Comment by glamarre360** [comment]
Yes that I understand but Imagine I have request1 that start an ExecutorService for request2. So you say those two request are reported independently. Then how do I know that in fact request2 has been lauched by request1 and that they're not two seperate requests.

**Comment by felixbarny** [comment]
The first part is TracingUtils.tracedExecutor which propagates the `SpanContext` of request1 to the thread which executes thread2. In thread2, a span is created like that:

```java
Span spanForRequest2 = tracer.buildSpan(methodSignature)
		.asChildOf(TracingUtils.getTraceContext().getCurrentSpan())
		.start();
```

The span for request2 is now associated with the same traceId and its parentId is the spanId of request1.

**Comment by glamarre360** [comment]
Ok I understand now. Would there be a way that the execution time of request1 take count of the time took by request 2 in his own execution time. That way we would know how much time the whole request took..

**Comment by felixbarny** [comment]
It's all about the latency the user perceives, right? If request1 blocks until request2 is finished, the execution time of request1 already contains the total execution time.

```
|-req1--------------------------|
                                      |-req2--|
                                      |-req3--|
```

If request1 triggers a async execution which does not block, the perceived latency for the user is only equal to the execution time of request1. 

```
|-req1--------------------------|
                                          |-req2---------------------|
```

**Comment by glamarre360** [comment]
That's true but in my case request1 reply to the client when the request2 returns with a callback so it will be
```
|-req1--------------------|                                            |-req1---|
                                       |-req2------------------------|
```

**Comment by felixbarny** [comment]
Wouldn't the callback be executed in yet another span like so:

```
|-req1--------------------|
                          |-req2------------------------|
                                                        |-req3---|
```

I'm not into the internals of Vert.x but if they offer something like a servlet filter it might be possible to set the total execution time as a custom tag on the last span (req3).

**Comment by glamarre360** [comment]
I don't consider that the callback is another request because the server only reply to the client inside that callback so in terms of latency the callback is part of the request1 elapsed time and the request 2 is also a part of it. So the complete latency perceived by the client is the elapsed time of request1, request 2 and the callback.
